x86/proofs/mldsa_poly_reduce.ml

```diff
@@ -0,0 +1,758 @@
+
+  (*** Now the result is just a replicated instance of our lemma ***)
+
+  CONV_TAC(EXPAND_CASES_CONV THENC ONCE_DEPTH_CONV NUM_MULT_CONV) THEN
+  ASM_REWRITE_TAC[WORD_ADD_0] THEN
+  DISCARD_STATE_TAC "s198" THEN
+  REWRITE_TAC[GSYM mldsa_barred]);;
+
+let MLDSA_POLY_REDUCE_NOIBT_SUBROUTINE_CORRECT = prove
+ (`!a x pc stackpointer returnaddress.
+        aligned 32 a /\
+        nonoverlapping (word pc,LENGTH mldsa_poly_reduce_tmc) (a,1024) /\
+        nonoverlapping (stackpointer,8) (a,1024)
+        ==> ensures x86
+             (\s. bytes_loaded s (word pc) mldsa_poly_reduce_tmc /\
+                  read RIP s = word pc /\
+                  read RSP s = stackpointer /\
+                  read (memory :> bytes64 stackpointer) s = returnaddress /\
+                  C_ARGUMENTS [a] s /\
+                  !i. i < 256
+                      ==> read(memory :> bytes32(word_add a (word(4 * i)))) s =
+                          x i)
+             (\s. read RIP s = returnaddress /\
+                  read RSP s = word_add stackpointer (word 8) /\
+                  !i. i < 256
+                      ==> ival(read(memory :> bytes32
+                                 (word_add a (word(4 * i)))) s) =
+                          ival(mldsa_barred (x i)))
+             (MAYCHANGE [RSP] ,, MAYCHANGE_REGS_AND_FLAGS_PERMITTED_BY_ABI ,,
+              MAYCHANGE [memory :> bytes(a,1024)])`,
+  X86_PROMOTE_RETURN_NOSTACK_TAC mldsa_poly_reduce_tmc MLDSA_REDUCE_CORRECT);;
+
+let MLDSA_POLY_REDUCE_SUBROUTINE_CORRECT = prove
+ (`!a x pc stackpointer returnaddress.
+        aligned 32 a /\
+        nonoverlapping (word pc,LENGTH mldsa_poly_reduce_mc) (a,1024) /\
+        nonoverlapping (stackpointer,8) (a,1024)
+        ==> ensures x86
+             (\s. bytes_loaded s (word pc) mldsa_poly_reduce_mc /\
+                  read RIP s = word pc /\
+                  read RSP s = stackpointer /\
+                  read (memory :> bytes64 stackpointer) s = returnaddress /\
+                  C_ARGUMENTS [a] s /\
+                  !i. i < 256
+                      ==> read(memory :> bytes32(word_add a (word(4 * i)))) s =
+                          x i)
+             (\s. read RIP s = returnaddress /\
+                  read RSP s = word_add stackpointer (word 8) /\
+                  !i. i < 256
+                      ==> ival(read(memory :> bytes32
+                                 (word_add a (word(4 * i)))) s) =
+                          ival(mldsa_barred (x i)))
+             (MAYCHANGE [RSP] ,, MAYCHANGE_REGS_AND_FLAGS_PERMITTED_BY_ABI ,,
+              MAYCHANGE [memory :> bytes(a,1024)])`,
+  MATCH_ACCEPT_TAC(ADD_IBT_RULE MLDSA_POLY_REDUCE_NOIBT_SUBROUTINE_CORRECT));;
+
+(* ------------------------------------------------------------------------- *)
+(* General lemmas to eventually put elsewhere.                               *)
+(* ------------------------------------------------------------------------- *)
```

Comment:
I can move these to a util/common location if there is a good place you have in mind

Reply:
They are pretty x86-specific, so probably a good place would be in `x86/proofs/x86.ml`. Perhaps just after `WINDOWS_MAYCHANGE_REGS_AND_FLAGS_PERMITTED_BY_ABI`, since they are concerned with `MAYCHANGE` and are intended to handle issues around the Windows ABI, even though they are not directly connected. In the fairly near future I'll wrap these up in some proper automation so we won't need to invoke that sort of gruesome proof manually every time, but we might need to try a few more examples first to see what the general case of a Windows wrapper looks like in the case of AVX2 usage.